File: bot/src/application/scouting/commands/scout_markets.py
"""Scout markets command and handler"""
from dataclasses import dataclass
from typing import List, Dict
import logging
import uuid
import asyncio
from pymediatr import Request, RequestHandler

from domain.shared.exceptions import ShipNotFoundError
from ports.repositories import IShipRepository

logger = logging.getLogger(__name__)

# Global lock to prevent race conditions in concurrent scout markets calls
_scout_markets_lock = asyncio.Lock()


@dataclass(frozen=True)
class ScoutMarketsResult:
    """Result of scout markets deployment"""
    container_ids: List[str]
    assignments: Dict[str, List[str]]  # ship_symbol -> assigned_markets
    reused_containers: List[str] = None  # Container IDs that were reused

    def __post_init__(self):
        """Set default for reused_containers"""
        if self.reused_containers is None:
            object.__setattr__(self, 'reused_containers', [])


@dataclass(frozen=True)
class ScoutMarketsCommand(Request[ScoutMarketsResult]):
    """
    Command to partition markets across multiple ships and deploy market scouting.

    The handler orchestrates:
    1. Load ships and validate they exist
    2. Get system graph from graph provider
    3. Call routing engine's optimize_fleet_tour() to partition markets (VRP)
    4. Create N ScoutTourCommand containers via daemon client (one per ship)
    5. Return container IDs and market assignments

    Each ship gets a disjoint subset of markets optimized by the VRP solver.
    Tours always return to start for multi-market assignments (2+).
    """
    ship_symbols: List[str]
    player_id: int
    system: str
    markets: List[str]
    iterations: int = 1


class ScoutMarketsHandler(RequestHandler[ScoutMarketsCommand, ScoutMarketsResult]):
    """
    Handler for scout markets deployment.

    Partitions markets across multiple ships using VRP and launches
    individual scout tour containers for each ship.
    """

    def __init__(self, ship_repository: IShipRepository):
        """
        Initialize ScoutMarketsHandler.

        Args:
            ship_repository: Repository for ship persistence
        """
        self._ship_repo = ship_repository

    async def handle(self, request: ScoutMarketsCommand) -> ScoutMarketsResult:
        """
        Partition markets and deploy market scouting with idempotency.

        Process:
        1. Check for existing active scout-tour containers for requested ships
        2. Reuse containers where possible (same ship, compatible parameters)
        3. Load ships and run VRP for ships needing new containers
        4. Create new scout-tour containers only for ships without active ones
        5. Return all container IDs (reused + newly created)

        Args:
            request: Scout markets command

        Returns:
            ScoutMarketsResult with container IDs, assignments, and reused containers

        Raises:
            ShipNotFoundError: If any ship doesn't exist
        """
        logger.info(f"Deploying scout markets: {len(request.ship_symbols)} ships, {len(request.markets)} markets")

        # Use lock to prevent race conditions in concurrent calls
        # This ensures only one scout markets call executes at a time
        async with _scout_markets_lock:
            # 1. Query existing active containers for these ships
            from configuration.container import get_daemon_client
            daemon = get_daemon_client()

            existing_containers = daemon.list_containers(player_id=request.player_id)
            existing_ship_containers = self._find_existing_scout_containers(
                existing_containers.get('containers', []),
                request.ship_symbols
            )

            logger.info(f"Found {len(existing_ship_containers)} existing containers to reuse")

            # 2. Partition ships: reuse vs need VRP
            ships_with_containers = set(existing_ship_containers.keys())
            ships_needing_containers = [s for s in request.ship_symbols if s not in ships_with_containers]

            reused_container_ids = list(existing_ship_containers.values())
            container_ids = reused_container_ids.copy()

            logger.info(f"Reusing {len(reused_container_ids)} containers, creating {len(ships_needing_containers)} new")

            # 3. If no ships need containers, return early with reused ones
            if not ships_needing_containers:
                logger.info("All ships have existing containers - returning reused containers")
                return ScoutMarketsResult(
                    container_ids=container_ids,
                    assignments={ship: [] for ship in request.ship_symbols},
                    reused_containers=reused_container_ids
                )

            # 4. Load ships needing containers and run VRP
            ships = []
            ship_locations = {}

            for ship_symbol in ships_needing_containers:
                ship = self._ship_repo.find_by_symbol(ship_symbol, request.player_id)
                if ship is None:
                    raise ShipNotFoundError(
                        f"Ship {ship_symbol} not found for player {request.player_id}"
                    )
                ships.append(ship)
                ship_locations[ship_symbol] = ship.current_location.symbol

            logger.info(f"Loaded {len(ships)} ships for VRP: {ship_locations}")

            # 5. Get system graph
            from configuration.container import get_graph_provider_for_player
            graph_provider = get_graph_provider_for_player(request.player_id)
            graph_result = graph_provider.get_graph(request.system)

            if not graph_result:
                raise Exception(f"Failed to get graph for system {request.system}")

            graph_data = graph_result.graph if hasattr(graph_result, 'graph') else graph_result

            if 'waypoints' not in graph_data:
                raise Exception(f"Graph for system {request.system} has no waypoints")

            # Convert graph result to Dict[str, Waypoint]
            from domain.shared.value_objects import Waypoint
            graph = {}
            for wp_symbol, wp_data in graph_data['waypoints'].items():
                graph[wp_symbol] = Waypoint(
                    symbol=wp_symbol,
                    waypoint_type=wp_data.get('type', 'UNKNOWN'),
                    x=wp_data.get('x', 0),
                    y=wp_data.get('y', 0),
                    system_symbol=request.system,
                    traits=tuple(wp_data.get('traits', [])),
                    has_fuel=wp_data.get('has_fuel', False)
                )

            # 6. Run VRP for ships needing containers
            if len(ships) == 1:
                # Single ship gets all markets
                single_ship = ships_needing_containers[0]
                assignments = {single_ship: request.markets}
                logger.info(f"Single ship optimization: {single_ship} assigned all {len(request.markets)} markets")
            else:
                # Multiple ships - use VRP
                from configuration.container import get_routing_engine
                routing_engine = get_routing_engine()

                fuel_capacity = ships[0].fuel_capacity
                engine_speed = ships[0].engine_speed

                loop = asyncio.get_event_loop()
                assignments = await loop.run_in_executor(
                    None,
                    routing_engine.optimize_fleet_tour,
                    graph,
                    request.markets,
                    ship_locations,
                    fuel_capacity,
                    engine_speed
                )

                if not assignments:
                    raise Exception("Fleet tour optimization failed")

                logger.info(f"VRP partitioning complete:")
                for ship, markets in assignments.items():
                    logger.info(f"  {ship}: {len(markets)} markets")

            # 7. Create scout tour containers for ships needing them
            for ship_symbol in ships_needing_containers:
                assigned_markets = assignments.get(ship_symbol, [])

                if not assigned_markets:
                    logger.warning(f"Ship {ship_symbol} has no markets assigned, skipping container creation")
                    continue

                # Generate unique container ID
                container_id = f"scout-tour-{ship_symbol.lower()}-{uuid.uuid4().hex[:8]}"

                # Create container for this ship's tour
                daemon.create_container({
                    'container_id': container_id,
                    'player_id': request.player_id,
                    'container_type': 'command',
                    'config': {
                        'command_type': 'ScoutTourCommand',
                        'params': {
                            'ship_symbols': [ship_symbol],  # Match expected format in tests
                            'ship_symbol': ship_symbol,
                            'player_id': request.player_id,
                            'system': request.system,
                            'markets': assigned_markets
                        },
                        'iterations': request.iterations
                    },
                    'restart_policy': 'no'
                })

                container_ids.append(container_id)
                logger.info(f"✅ Created container {container_id} for {ship_symbol} ({len(assigned_markets)} markets)")

            # Build full assignments dict (include ships with reused containers)
            full_assignments = {ship: [] for ship in request.ship_symbols}
            full_assignments.update(assignments)

            logger.info(f"Scout markets deployment complete: {len(container_ids)} total containers " +
                       f"({len(reused_container_ids)} reused, {len(container_ids) - len(reused_container_ids)} created)")

            return ScoutMarketsResult(
                container_ids=container_ids,
                assignments=full_assignments,
                reused_containers=reused_container_ids
            )

    def _find_existing_scout_containers(
        self,
        containers: List[Dict],
        requested_ships: List[str]
    ) -> Dict[str, str]:
        """
        Find existing active scout-tour containers for requested ships.

        Args:
            containers: List of container dicts from daemon
            requested_ships: Ship symbols we're checking for

        Returns:
            Dict mapping ship_symbol -> container_id for ships with active containers
        """
        existing = {}

        for container_info in containers:
            container_id = container_info['container_id']
            status = container_info['status']

            # Only consider STARTING or RUNNING containers
            if status not in ['STARTING', 'RUNNING']:
                continue

            # Parse ship symbol from scout-tour container ID pattern
            if container_id.startswith('scout-tour-'):
                # Extract ship symbol (everything between 'scout-tour-' and last '-')
                parts = container_id[len('scout-tour-'):].rsplit('-', 1)
                if len(parts) == 2:
                    ship_from_container = parts[0].upper()

                    # Check if this ship is in our requested list
                    if ship_from_container in requested_ships:
                        # Store first active container for each ship
                        if ship_from_container not in existing:
                            existing[ship_from_container] = container_id
                            logger.info(f"Found existing container {container_id} for {ship_from_container} (status: {status})")

        return existing

File: gobot/internal/application/scouting/scout_markets.go
package scouting

import (
	"context"
	"fmt"
	"strings"

	"github.com/google/uuid"

	"github.com/andrescamacho/spacetraders-go/internal/application/common"
	"github.com/andrescamacho/spacetraders-go/internal/domain/daemon"
	"github.com/andrescamacho/spacetraders-go/internal/domain/navigation"
	"github.com/andrescamacho/spacetraders-go/internal/domain/routing"
	"github.com/andrescamacho/spacetraders-go/internal/domain/system"
)

// ScoutMarketsCommand orchestrates fleet deployment for market scouting
// Uses VRP optimization to distribute markets across multiple ships
// Idempotent: reuses existing containers for ships that already have them
type ScoutMarketsCommand struct {
	PlayerID     uint
	ShipSymbols  []string
	SystemSymbol string
	Markets      []string
	Iterations   int // Number of iterations (-1 for infinite)
}

// ScoutMarketsResponse contains container IDs and market assignments
type ScoutMarketsResponse struct {
	ContainerIDs     []string            // All container IDs (new + reused)
	Assignments      map[string][]string // ship_symbol -> markets assigned
	ReusedContainers []string            // Subset of ContainerIDs that were reused
}

// ScoutMarketsHandler handles the scout markets command
type ScoutMarketsHandler struct {
	shipRepo               navigation.ShipRepository
	graphProvider          system.ISystemGraphProvider
	routingClient          routing.RoutingClient
	daemonClient           daemon.DaemonClient
	shipAssignmentRepo     daemon.ShipAssignmentRepository
}

// NewScoutMarketsHandler creates a new scout markets handler
func NewScoutMarketsHandler(
	shipRepo navigation.ShipRepository,
	graphProvider system.ISystemGraphProvider,
	routingClient routing.RoutingClient,
	daemonClient daemon.DaemonClient,
	shipAssignmentRepo daemon.ShipAssignmentRepository,
) *ScoutMarketsHandler {
	return &ScoutMarketsHandler{
		shipRepo:               shipRepo,
		graphProvider:          graphProvider,
		routingClient:          routingClient,
		daemonClient:           daemonClient,
		shipAssignmentRepo:     shipAssignmentRepo,
	}
}

// Handle executes the scout markets command
func (h *ScoutMarketsHandler) Handle(ctx context.Context, request common.Request) (common.Response, error) {
	cmd, ok := request.(*ScoutMarketsCommand)
	if !ok {
		return nil, fmt.Errorf("invalid request type")
	}

	// 0. Stop existing scout-tour containers for the ships in this command
	// This ensures VRP recalculation with fresh ship positions
	// User explicitly ran scout-all-markets, so we want to redistribute work
	for _, shipSymbol := range cmd.ShipSymbols {
		// Check if ship has an active assignment
		assignment, err := h.shipAssignmentRepo.FindByShip(ctx, shipSymbol, int(cmd.PlayerID))
		if err != nil {
			return nil, fmt.Errorf("failed to query ship assignment for %s: %w", shipSymbol, err)
		}

		if assignment != nil && assignment.Status() == "active" {
			// Stop the container and release the ship
			containerID := assignment.ContainerID()
			fmt.Printf("[ScoutMarkets] Stopping existing container %s for ship %s (scout-all-markets reset)\n", containerID, shipSymbol)

			if err := h.daemonClient.StopContainer(ctx, containerID); err != nil {
				// Non-fatal: container might already be stopped or not found
				fmt.Printf("[ScoutMarkets] Warning: Failed to stop container %s: %v\n", containerID, err)
			}

			// Release ship assignment
			if err := h.shipAssignmentRepo.Release(ctx, shipSymbol, int(cmd.PlayerID), "scout_all_markets_reset"); err != nil {
				// Non-fatal: assignment might already be released
				fmt.Printf("[ScoutMarkets] Warning: Failed to release ship %s: %v\n", shipSymbol, err)
			}
		}
	}

	// 1. Query ship assignments to find existing active assignments (source of truth)
	// After cleanup above, this should find no active assignments for the requested ships
	shipsWithContainers := make(map[string]string) // ship -> container_id
	reusedContainers := []string{}

	for _, shipSymbol := range cmd.ShipSymbols {
		assignment, err := h.shipAssignmentRepo.FindByShip(ctx, shipSymbol, int(cmd.PlayerID))
		if err != nil {
			return nil, fmt.Errorf("failed to query ship assignment for %s: %w", shipSymbol, err)
		}

		if assignment != nil && assignment.Status() == "active" {
			shipsWithContainers[shipSymbol] = assignment.ContainerID()
			reusedContainers = append(reusedContainers, assignment.ContainerID())
		}
	}

	// 2. Partition ships: with_containers vs needing_containers
	shipsNeedingContainers := []string{}
	for _, ship := range cmd.ShipSymbols {
		if _, exists := shipsWithContainers[ship]; !exists {
			shipsNeedingContainers = append(shipsNeedingContainers, ship)
		}
	}

	// 3. Early return if all ships have containers
	if len(shipsNeedingContainers) == 0 {
		allContainerIDs := []string{}
		assignments := make(map[string][]string)
		for ship, containerID := range shipsWithContainers {
			allContainerIDs = append(allContainerIDs, containerID)
			assignments[ship] = []string{} // Unknown assignments for reused
		}

		return &ScoutMarketsResponse{
			ContainerIDs:     allContainerIDs,
			Assignments:      assignments,
			ReusedContainers: reusedContainers,
		}, nil
	}

	// 4. Load ships and get current locations + specs
	shipConfigs := make(map[string]*routing.ShipConfigData)

	for _, shipSymbol := range shipsNeedingContainers {
		shipData, err := h.shipRepo.FindBySymbol(ctx, shipSymbol, int(cmd.PlayerID))
		if err != nil {
			return nil, fmt.Errorf("failed to load ship %s: %w", shipSymbol, err)
		}

		shipConfigs[shipSymbol] = &routing.ShipConfigData{
			CurrentLocation: shipData.CurrentLocation().Symbol,
			FuelCapacity:    shipData.FuelCapacity(),
			EngineSpeed:     shipData.EngineSpeed(),
		}
	}

	// 5. Get system graph
	graphResult, err := h.graphProvider.GetGraph(ctx, cmd.SystemSymbol, false, int(cmd.PlayerID))
	if err != nil {
		return nil, fmt.Errorf("failed to get graph: %w", err)
	}

	// Convert graph to waypoint data
	waypointData, err := extractWaypointData(graphResult.Graph)
	if err != nil {
		return nil, fmt.Errorf("failed to extract waypoint data: %w", err)
	}

	// 6. Run VRP optimization
	fmt.Printf("[DEBUG ScoutMarkets] Ships needing containers: %d, Markets: %v\n", len(shipsNeedingContainers), cmd.Markets)
	var assignments map[string][]string
	if len(shipsNeedingContainers) == 1 {
		// Single ship: assign all markets
		fmt.Printf("[DEBUG ScoutMarkets] Single ship path - assigning all markets to %s\n", shipsNeedingContainers[0])
		assignments = map[string][]string{
			shipsNeedingContainers[0]: cmd.Markets,
		}
	} else {
		// Multi-ship: use VRP via routing client
		fmt.Printf("[DEBUG ScoutMarkets] Multi-ship path - calling VRP for %d ships and %d markets\n", len(shipsNeedingContainers), len(cmd.Markets))
		vrpRequest := &routing.VRPRequest{
			SystemSymbol:    cmd.SystemSymbol,
			ShipSymbols:     shipsNeedingContainers,
			MarketWaypoints: cmd.Markets,
			ShipConfigs:     shipConfigs,
			AllWaypoints:    waypointData,
		}

		vrpResponse, err := h.routingClient.PartitionFleet(ctx, vrpRequest)
		if err != nil {
			return nil, fmt.Errorf("VRP optimization failed: %w", err)
		}

		// Extract market assignments from VRP response
		assignments = make(map[string][]string)
		for shipSymbol, tourData := range vrpResponse.Assignments {
			fmt.Printf("[DEBUG ScoutMarkets] VRP assigned %d markets to %s: %v\n", len(tourData.Waypoints), shipSymbol, tourData.Waypoints)
			assignments[shipSymbol] = tourData.Waypoints
		}
	}

	// 7. Create scout-tour containers for ships needing them
	newContainerIDs := []string{}
	for shipSymbol, markets := range assignments {
		containerID := fmt.Sprintf("scout-tour-%s-%s",
			strings.ToLower(shipSymbol),
			generateShortUUID())

		scoutTourCmd := &ScoutTourCommand{
			PlayerID:   cmd.PlayerID,
			ShipSymbol: shipSymbol,
			Markets:    markets,
			Iterations: cmd.Iterations,
		}

		err := h.daemonClient.CreateScoutTourContainer(ctx, containerID, cmd.PlayerID, scoutTourCmd)
		if err != nil {
			return nil, fmt.Errorf("failed to create container for %s: %w", shipSymbol, err)
		}

		// Ship assignment is now automatically created by ContainerRunner.Start()
		// based on the "ship_symbol" config value

		newContainerIDs = append(newContainerIDs, containerID)
	}

	// 8. Combine results
	allContainerIDs := append(reusedContainers, newContainerIDs...)

	// Add reused containers to assignments (with empty markets list)
	for ship := range shipsWithContainers {
		if _, exists := assignments[ship]; !exists {
			assignments[ship] = []string{}
		}
	}

	return &ScoutMarketsResponse{
		ContainerIDs:     allContainerIDs,
		Assignments:      assignments,
		ReusedContainers: reusedContainers,
	}, nil
}

// generateShortUUID generates an 8-character hex UUID
func generateShortUUID() string {
	id := uuid.New()
	// Use first 8 characters of the UUID string (without hyphens)
	return strings.ReplaceAll(id.String(), "-", "")[:8]
}

// extractWaypointData converts graph format to routing waypoint data
func extractWaypointData(graph map[string]interface{}) ([]*system.WaypointData, error) {
	waypoints, ok := graph["waypoints"].(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("invalid graph format: missing waypoints")
	}

	waypointData := make([]*system.WaypointData, 0, len(waypoints))
	for symbol, data := range waypoints {
		wpMap, ok := data.(map[string]interface{})
		if !ok {
			continue
		}

		x, _ := wpMap["x"].(float64)
		y, _ := wpMap["y"].(float64)
		hasFuel, _ := wpMap["has_fuel"].(bool)

		waypointData = append(waypointData, &system.WaypointData{
			Symbol:  symbol,
			X:       x,
			Y:       y,
			HasFuel: hasFuel,
		})
	}

	return waypointData, nil
}
